src/Stratis.Bitcoin/Consensus/ConsensusManager.cs

```diff
@@ -414,67 +414,89 @@ private async Task OnPartialValidationSucceededAsync(ChainedHeader chainedHeader
+                        shouldPartialValidationContinue = true;
                 }
 
-                this.logger.LogTrace("Full validation is{0} required.", fullValidationRequired ? "" : " NOT");
+                this.logger.LogTrace("Full validation is{0} required.", partialResult.FullValidationRequired ? string.Empty : " NOT");
 
-                if (fullValidationRequired)
+                if (partialResult.FullValidationRequired)
                 {
-                    connectBlocksResult = await this.FullyValidateLockedAsync(chainedHeader).ConfigureAwait(false);
+                    fullResult = await this.FullyValidateAsync(chainedHeader).ConfigureAwait(false);
+                    shouldPartialValidationContinue = fullResult.Succeeded;
                 }
             }
 
-            if (connectBlocksResult != null)
+            // If there was no full validation, partial validation should continue.
+            // If there was full validation, partial validation should continue only if full validation succeeded.
+            if (shouldPartialValidationContinue)
             {
-                if (connectBlocksResult.PeersToBan != null)
-                {
-                    var peersToBan = new List<INetworkPeer>();
+                this.logger.LogTrace("Partial validation of {0} block will be started.", partialResult.ChainedHeaderBlocksToValidate.Count);
 
-                    lock (this.peerLock)
-                    {
-                        foreach (int peerId in connectBlocksResult.PeersToBan)
-                        {
-                            if (this.peersByPeerId.TryGetValue(peerId, out INetworkPeer peer))
-                                peersToBan.Add(peer);
-                        }
-                    }
+                // Start validating all next blocks that come after the current block,
+                // all headers in this list have the blocks present in the header.
+                foreach (ChainedHeaderBlock toValidate in partialResult.ChainedHeaderBlocksToValidate)
+                    this.partialValidator.StartPartialValidation(toValidate.ChainedHeader, toValidate.Block, this.OnPartialValidationCompletedCallbackAsync);
+            }
 
-                    this.logger.LogTrace("{0} peers will be banned.", peersToBan.Count);
+            this.logger.LogTrace("(-)");
+        }
 
-                    foreach (INetworkPeer peer in peersToBan)
-                        this.peerBanning.BanAndDisconnectPeer(peer.PeerEndPoint, connectBlocksResult.BanDurationSeconds, connectBlocksResult.BanReason);
-                }
+        /// <summary>
+        /// Fully validates the chained header and should it fail, ban peers that has this header.
+        /// </summary>
+        /// <param name="chainedHeader">The chained header to validate.</param>
+        /// <returns>Validation related information.</returns>
+        private async Task<ConnectBlocksResult> FullyValidateAsync(ChainedHeader chainedHeader)
```

Comment:
@Aprogiena this is just a copy paste of the peer banning code in the previous version of the code. No logic changes.

Reply:
You need to add `Locked` on this method as its under a lock.

Reply:
OK will fix now 👍